### PR TITLE
KIALI-949 Update release instructions

### PR DESCRIPTION
**Pull request by cfcosta**
This is the core complement of https://github.com/kiali/kiali-ui/pull/516, and details all our release process.

### CODE REVIEW

File: RELEASING.adoc [hunk 1]
```diff
@@ -1,80 +1,177 @@
...
+  * Istio Ansible Installer
+  * Openshift Istio Ansible Installer (istiooc)
+* Update JIRA issues included with the release
+* Send out notifications (twitter, mailing lists, IRC)
 
-The steps to perform a release are relatively simple thanks to the way link:.travis.yml[Travis] is setup.
+== Creating a new JIRA issue
 
-Before running these steps, it is assumed the link:https://github.com/kiali/kiali/tree/master[kiali master] branch and the link:https://github.com/kiali/kiali-ui/tree/master[kiali-ui master] branch are all up to date and include all code that is to be released and that your local `master` branches are up to date with the remote `master` branches.
+Create a new issue on link:https://issues.jboss.org[JIRA] detailing the
```
Comment (jmazzitelli) [inline RELEASING.adoc]: JIRA link should go to the KIALI project: https://issues.jboss.org/browse/KIALI

File: RELEASING.adoc [hunk 2]
```diff
@@ -1,80 +1,177 @@
...
+* Make sure Getting Started docs are complete
+* Update Kiali UI version
+* Publish the API Docs
+* Update Makefile
+* Create the Release PR
+* Tag Git Repo
+* Notify Istio and Openshift Istio Teams for changes:
+  * Istio Helm Charts
+  * Istio Ansible Installer
+  * Openshift Istio Ansible Installer (istiooc)
```
Comment (jmazzitelli) [inline RELEASING.adoc]: The asciidoc preview  is showing these three lines above are not being listed as sub-bullets, but rather on the same level as the "Notify Istio..." line. I don't know why - this looks correct with the two-space indentation. Keep an eye on this when we merge - not sure if the preview is just wrong or if something really is wrong with the asciidoc here.

File: RELEASING.adoc [hunk 3]
```diff
@@ -1,80 +1,177 @@
...
-1) Tag the Kiali-UI master branch in this repo: https://github.com/kiali/kiali-ui
+The docs can be found link:https://github.com/kiali/kiali.io[here]. Do the
+changes, and create a PR.
 
-NOTE: You must have write permissions to the public GitHub repository in order to be able to do this.
+== Updating Kiali UI
 
-For example,
+The specifics of releasing Kiali UI can be found
+link:https://github.com/kiali/kiali-ui/blob/master/RELEASING.adoc[here].
```
Comment (jmazzitelli) [inline RELEASING.adoc]: This document doesn't exist, so this link results in a 404.

File: RELEASING.adoc [hunk 4]
```diff
@@ -1,80 +1,177 @@
...
+
+Once the tag has been pushed, Travis will immediately run another build, but
+this time will tag the docker image with the same version string as the git tag
+name. In the example above, the docker image pushed to DockerHub will not be
+tagged with "latest" but will instead be tagged with "v1.0.0".
+
+You can let the committers know that the release has been completed and PRs can
+begin to be merged again. Though, it would be best to wait for this
+notification to go out until after you bump up the version string to the next
+snapshot version (see "Prepare For The Next Release" below).
```
Comment (jmazzitelli) [inline RELEASING.adoc]: Looks like this "Prepare For The Next Release" section has been deleted in this PR. I think we need to add that back - at the end of the release process, we have to tell the reader that they need to set the Makefile back to CONSOLE_VERSION=latest and VERSION=<the next version-SNAPSHOT> to get ready for the next iteration of development.

File: RELEASING.adoc [hunk 5]
```diff
@@ -1,80 +1,177 @@
...
+
+At this point, Travis will run the test suite, build the binary, and push the
+versioned docker image to link:https://hub.docker.com/r/kiali/kiali[DockerHub].
+In the example above, the docker image pushed to DockerHub will not be tagged
+with "latest" but will instead be tagged with "1.0.0.Final."
```
Comment (jmazzitelli) [inline RELEASING.adoc]: This paragraph ("At this point, ....") can be deleted, it is pretty much saying the same thing as the earlier paragraph ("Once the tag has been pushed,...").

File: RELEASING.adoc [hunk 6]
```diff
@@ -1,80 +1,177 @@
...
+diff --git a/Makefile b/Makefile
+index c1f7f80..1d05ff3 100644
+--- a/Makefile
++++ b/Makefile
+@@ -13,7 +13,7 @@ COMMIT_HASH ?= $(shell git rev-parse HEAD)
+ # WARNING: If you have previously run the 'docker' target but
+ # later want to change the CONSOLE_VERSION then you must run
+ # the 'clean' target first before re-running the 'docker' target.
+-CONSOLE_VERSION ?= latest
++CONSOLE_VERSION ?= v1.0.0
```
Comment (jmazzitelli) [inline RELEASING.adoc]: Today, the  CONSOLE_VERSION is very different than the VERSION. We should tell people this CONSOLE_VERSION comes from the npm repository and may not be (in fact, today it is not) the same as VERSION. This is the npm repo link where you get the console version:

https://www.npmjs.com/package/@kiali/kiali-ui?activeTab=versions

File: RELEASING.adoc (hunk 2 above)
Comment (cfcosta) [reply RELEASING.adoc]: I think the syntax for asciidoc for listings is different from markdown. I'll check.

File: RELEASING.adoc (hunk 3 above)
Comment (cfcosta) [reply RELEASING.adoc]: It will exist once https://github.com/kiali/kiali-ui/pull/516 gets merged.

File: RELEASING.adoc (hunk 2 above)
Comment (JStickler) [reply RELEASING.adoc]: In asciidoc a blank line is required before and after a list to separated it from other blocks.

File: RELEASING.adoc (hunk 5 above)
Comment (cfcosta) [reply RELEASING.adoc]: That was actually copied from what you did write :P

File: RELEASING.adoc (hunk 4 above)
Comment (cfcosta) [reply RELEASING.adoc]: Done, added it back.

File: RELEASING.adoc (hunk 6 above)
Comment (cfcosta) [reply RELEASING.adoc]: True, added an addendum on it.

File: RELEASING.adoc [hunk 7]
```diff
@@ -1,80 +1,210 @@
...
 
-Before running these steps, it is assumed the link:https://github.com/kiali/kiali/tree/master[kiali master] branch and the link:https://github.com/kiali/kiali-ui/tree/master[kiali-ui master] branch are all up to date and include all code that is to be released and that your local `master` branches are up to date with the remote `master` branches.
+Create a new issue on link:https://issues.jboss.org/browse/KIALI[JIRA] detailing the
+release, and add sub-tasks if necessary.
 
-For the steps below, it is assumed the image version being released is v1.0.0. Obviously, this version string for the docker image will change as new images are released.
+== Updating Docs
 
-== Kiali-UI
+First step is to make sure the "Getting Started" document for the website is up
```
Comment (JStickler) [inline RELEASING.adoc]: The next step (creating the JIRA is the first step).

File: RELEASING.adoc [hunk 8]
```diff
@@ -1,80 +1,210 @@
...
+to date and works with the new version.
 
-1) Tag the Kiali-UI master branch in this repo: https://github.com/kiali/kiali-ui
+The docs can be found link:https://github.com/kiali/kiali.io[here]. Do the
```
Comment (JStickler) [inline RELEASING.adoc]: Update the document and create a PR.

File: RELEASING.adoc [hunk 9]
```diff
@@ -1,80 +1,210 @@
...
 
-For example,
+The specifics of releasing Kiali UI can be found
+link:https://github.com/kiali/kiali-ui/blob/master/RELEASING.adoc[here].
 
-```
-$ git tag v1.0.0
-$ git push origin v1.0.0
-```
+Releasing Kiali UI should come before releasing the core, as the image is
```
Comment (JStickler) [inline RELEASING.adoc]: You should release Kiali UI before releasing Kiali core, as Travis will create the new image as soon as the tag for core is pushed to master.

File: RELEASING.adoc (hunk 5 above)
Comment (jmazzitelli) [reply RELEASING.adoc]: Yes, I remember saying I should delete it, but didn't want to write a separate PR just for that - so I wanted for something else to come along to piggy back on it :)

File: RELEASING.adoc [hunk 10]
```diff
@@ -1,80 +1,210 @@
...
+created as soon as the tag for core is pushed to master.
+
+== Publish API docs
+
+API docs are generated by swagger, and this comment will both install it and
```
Comment (JStickler) [inline RELEASING.adoc]: API docs are generated by Swagger, and the following command will both install Swagger and generate the updated documentation.

File: RELEASING.adoc [hunk 11]
```diff
@@ -1,80 +1,210 @@
...
+[source, bash]
+----
+$ make swagger-install swagger-gen
+----
 
-2) Go to https://www.npmjs.com/package/@kiali/kiali-ui and find the version of the UI that should be built for this release. We will put this console version number in Kiali's Makefile - see below.
+== Update Makefile
 
-_Finally, it is still undetermined how future releases are going to be done with respect to kiali-ui and its own version string. It is possible kiali-ui and kiali repos will be combined into one and thus both kiali and the kiali-ui version strings will be united as a single version string. Today, the version string of kiali-ui as it is seen in the NPM repository is determined within the kiali-ui travis configuration file. This may change in the future. But for now, we need to go to the NPM repository to get the version string of kiali-ui that is to be pulled into the kiali image._
+Change the VERSION variable on Makefile to the new version, this is what is
```
Comment (JStickler) [inline RELEASING.adoc]: Change the `VERSION` variable in the Makefile to the new version.  This version will be displayed in the UI, and is also what Travis will use to trigger the release process.

File: RELEASING.adoc [hunk 12]
```diff
@@ -1,80 +1,210 @@
...
++VERSION ?= v1.0.0
+ COMMIT_HASH ?= $(shell git rev-parse HEAD)
 
-* The VERSION variable should be set to the new version string of the upcoming release.
-* The CONSOLE_VERSION variable should be set to the kiali-ui npm version string corresponding to the console version to be used.
+ # Indicates which version of the UI console is to be embedded
+----
 
-For example,
+By default, the Makefile gets the latest version of the UI, so we'll also
```
Comment (JStickler) [inline RELEASING.adoc]: By default, the Makefile gets the latest version of the UI, so you'll also need to update the console version in the Makefile:

File: RELEASING.adoc [hunk 13]
```diff
@@ -1,80 +1,210 @@
...
-2) Commit that change to your private GitHub repository:
+ # Version label is used in the OpenShift/K8S resources to identify
+----
 
-```
-$ git add Makefile
-$ git commit -m "New release"
-$ git push <your remote repo name> master
-```
+One thing to keep in mind is that CONSOLE_VERSION comes from the npm
```
Comment (JStickler) [inline RELEASING.adoc]: Move this up before the code sample.

File: RELEASING.adoc [hunk 14]
```diff
@@ -1,80 +1,210 @@
...
+----
 $ git pull
-$ git tag v1.0.0
+$ git tag v1.0.0 <merge commit id hash>
 $ git push origin v1.0.0
-```
+----
+
+Once the tag has been pushed, Travis will immediately run another build, but
+this time will tag the docker image with the same version string as the git tag
```
Comment (JStickler) [inline RELEASING.adoc]: Docker
Git

File: RELEASING.adoc [hunk 15]
```diff
@@ -1,80 +1,210 @@
...
+diff --git a/Makefile b/Makefile
+index c1f7f80..4a8af40 100644
+--- a/Makefile
++++ b/Makefile
+@@ -3,7 +3,7 @@ SHELL=/bin/bash
+ 
+ # Identifies the current build.
+ # These will be embedded in the app and displayed when it starts.
+-VERSION ?= v1.0.0
++VERSION ?= v1.1.0
```
Comment (jmazzitelli) [inline RELEASING.adoc]: Do we want to add a suffix to the new development version as we've been doing, such as "-SNAPSHOT" so the version is `v1.1.0-SNAPSHOT` in this example. Anyone getting images showing this SNAPSHOT version in the About box and logs will see it is obviously a development build (whereas if someone see v1.1.0, it won't be obvious that this isn't a released image).

File: RELEASING.adoc [hunk 16]
```diff
@@ -1,80 +1,210 @@
...
+begin to be merged again. Though, it would be best to wait for this
+notification to go out until after you bump up the version string to the next
+snapshot version (see "Prepare For The Next Release" below).
+
+Note that, in the example above, the docker image pushed to DockerHub will not
+be tagged with "latest" but will instead be tagged with "1.0.0.Final."
+
+== Notify Istio and Openshift Istio Teams for changes
+
+There are no fixed steps for releasing a new version on the istio installer,
```
Comment (JStickler) [inline RELEASING.adoc]: Istio

**Comment by jmazzitelli** [comment]
Note Julie's comments/suggestions also. They should be incorporated.

File: RELEASING.adoc (hunk 15 above)
Comment (theute) [reply RELEASING.adoc]: +1 and I was wondering the same about the UI https://github.com/kiali/kiali-ui/pull/516

**Comment by theute** [comment]
Makes sense, I'm not sure we need to create a JIRA for the task.
We'll definitely need automation, to deal with JIRA, open the Pull Requests on the various repos...
Maybe something for the next sprint, but we need to agree how we want to trigger the build and then execute the various steps...
So agree on the tool language (go ? python ?) and/or if we build a bot to interact with the release and other tasks...

File: RELEASING.adoc [hunk 17]
```diff
@@ -1,80 +1,210 @@
 = Releasing Kiali
 
-This document gives the steps required to perform a release of Kiali. This document is only useful to Kiali release managers.
+This document covers the specifics of releasing Kiali, and is meant as a guide
+for Kiali release managers.
+
+The summary of the process is the following:
```
Comment (abonas) [inline RELEASING.adoc]: is the following -> maybe 'as follows' sounds better?

File: RELEASING.adoc [hunk 18]
```diff
@@ -1,80 +1,210 @@
...
+
+* Create a new JIRA issue for the release
```
Comment (abonas) [inline RELEASING.adoc]: JIRA issue of type...?

File: RELEASING.adoc [hunk 19]
```diff
@@ -1,80 +1,210 @@
...
+* Make sure Getting Started docs are complete
+* Update Kiali UI version
```
Comment (abonas) [inline RELEASING.adoc]: link to where and how to update?

File: RELEASING.adoc [hunk 20]
```diff
@@ -1,80 +1,210 @@
...
+* Publish the API Docs
+* Update Makefile
+* Create the Release PR
+* Tag Git Repo
```
Comment (abonas) [inline RELEASING.adoc]: how to tag/tagging strategy for numbers?

File: RELEASING.adoc [hunk 21]
```diff
@@ -1,80 +1,210 @@
...
+* Notify Istio and Openshift Istio Teams for changes:
```
Comment (abonas) [inline RELEASING.adoc]: I think it would be good to add links where/how to notify

File: RELEASING.adoc (hunk 19 above)
Comment (cfcosta) [reply RELEASING.adoc]: It's on the section, there's a link there to the kiali-ui release steps (which do not exist yet per https://github.com/kiali/kiali-ui/pull/516).

File: RELEASING.adoc (hunk 20 above)
Comment (cfcosta) [reply RELEASING.adoc]: It's on the section for it, this is just a summary.

File: RELEASING.adoc (hunk 21 above)
Comment (cfcosta) [reply RELEASING.adoc]: Same as the last comment, this is just a summary on what to do. The steps themselves are on the relevant section.

File: RELEASING.adoc (hunk 18 above)
Comment (cfcosta) [reply RELEASING.adoc]: What do you think it should be? There's just "Dev Task" and "Task" that apply here. Maybe "Release Task" should be a type on JIRA?

Comment (theute) [reply RELEASING.adoc]: Task would work if any.

**Comment by cfcosta** [comment]
@theute for sure, we've been talking about tackling automatization on Saturn. I think we could investigate using [homu](https://github.com/barosl/homu) for that, it's what the Rust team have been using, and it gives us more guarantees that we do not have right now, such as "Are the tests still passing when the PR is merged into master?", "When to merge a new PR?", and "What kind of acceptance criteria do we need for this change? Better test coverage?".

With some changes, it could also take care of closing the issues after they are merged, and moving them after they are released. I've been playing with this on a local jira, but for the one that we use, I don't think there are API Keys enabled for us to play with.

And about the JIRA issue, what do you think, maybe we shouldn't create a task just for the release and just make sure that everything that's "Ready for Release" is actually released?

**Comment by cfcosta** [comment]
Updated with review feedback.

**Pull request merged**